cinema/views.py

```diff
@@ -37,22 +44,103 @@ class MovieViewSet(viewsets.ModelViewSet):
     def get_serializer_class(self):
         if self.action == "list":
             return MovieListSerializer
-
         if self.action == "retrieve":
             return MovieDetailSerializer
-
         return MovieSerializer
 
+    def get_queryset(self):
+        queryset = Movie.objects.all().prefetch_related("actors", "genres")
+
+        actors = self.request.GET.get("actors")
+        genres = self.request.GET.get("genres")
+        title = self.request.GET.get("title")
+
+        if actors:
+            actors_ids = [int(actor_id) for actor_id in actors.split(",")]
+            queryset = queryset.filter(
+                actors__id__in=actors_ids).order_by("actors")
+
+        if genres:
+            genres_ids = [int(genre_id) for genre_id in genres.split(",")]
+            queryset = queryset.filter(
+                genres__id__in=genres_ids).order_by("genres")
+
+        if title:
+            queryset = queryset.filter(title__icontains=title)
+
+        return queryset.distinct()
+
 
 class MovieSessionViewSet(viewsets.ModelViewSet):
     queryset = MovieSession.objects.all()
     serializer_class = MovieSessionSerializer
 
     def get_serializer_class(self):
         if self.action == "list":
-            return MovieSessionListSerializer
-
+            return MovieSessionAdvancedListSerializer
         if self.action == "retrieve":
             return MovieSessionDetailSerializer
-
         return MovieSessionSerializer
+
+    def get_queryset(self):
+        queryset = MovieSession.objects.all().prefetch_related(
+            "movie",
+            "cinema_hall"
+        )
+
+        date = self.request.GET.get("date")
+        movies = self.request.GET.get("movie")
+
+        if self.action == "list":
+            queryset = (
+                queryset
+                .annotate(
+                    tickets_available=(
+                        F("cinema_hall__rows")
+                        * F("cinema_hall__seats_in_row")
+                        - Count("tickets"))
+                )
+            )
+
+        if date:
```

Comment:
Żeby uniknąć jednej duużej funkcji możemy podzielić ją na kilka mniejszych. Warto byłoby zdefiniować sobie funkcję prywatną 
```python
def _filter_by_date(self, queryset, date):
    try:
        start_date = make_aware(datetime.strptime(date, "%Y-%m-%d"))
        end_date = start_date + timedelta(days=1)
        return queryset.filter(show_time__gte=start_date, show_time__lt=end_date)
    except ValueError:
        raise ValidationError("Invalid date format. Use YYYY-MM-DD")
```
i potem tu w kodzie wywołać
```suggestion
        if date:
            queryset = self._filter_by_date(queryset, date)
```
Analogicznie byłoby z `movies` poniżej. To tylko taka sugestia na przyszłość.